File: src/events/check_run.rs
use crate::errors::Error;
use crate::events::Event;
use crate::events::base::{CheckRun, Repository, WebhookMessage};
use serde::Deserialize;

#[derive(Deserialize)]
pub struct CheckRunEvent {
    pub action: String,
    pub check_run: CheckRun,
    pub repository: Repository,
}

impl Event for CheckRunEvent {
    fn handle(&self) -> Result<Option<WebhookMessage>, Error> {
        if self.action != "completed" {
            return Ok(None);
        }

        if self.repository.private {
            return Ok(None);
        }

        if let Some(conclusion) = &self.check_run.conclusion {
            if self.check_run.app.slug != "cloudflare-workers-and-pages" {
                return Ok(None);
            }

            let emoji = if conclusion == "failure" {
                std::env::var("FAILURE_EMOJI").unwrap_or_else(|_| "❌".into())
            } else {
                std::env::var("SUCCESS_EMOJI").unwrap_or_else(|_| "✅".into())
            };

            let branch_name = self.check_run
                .check_suite
                .head_branch
                .as_deref()
                .unwrap_or("unknown");
            let formatted = format!(
                "{emoji} Check [{conclusion}](<{}>) on [{}](<{}>)/[{branch_name}](<{}/tree/{branch_name}>)",
                self.check_run.html_url,
                self.repository.name,
                self.repository.html_url,
                self.repository.html_url,
            );

            return Ok(Some(WebhookMessage {
                content: formatted,
                username: "Cloudflare Pages".to_string(),
                avatar_url: self.check_run.app.owner.avatar_url.clone(),
            }));
        }

        Ok(None)
    }
}

File: internal/events/checkRun.go
package events

import (
	"fmt"
	"gohook/internal/config"
	"gohook/internal/structs/discord"
	"gohook/internal/structs/github"
	"slices"
)

var ignoredChecks = []string{
	"Dependabot",
	"GitHub Actions",
	"GitHub Advanced Security",
}

type CheckRun struct {
	Action   string `json:"action"`
	CheckRun struct {
		Conclusion string `json:"conclusion"`
		HtmlUrl    string `json:"html_url"`
		App        struct {
			Name string `json:"name"`
		} `json:"app"`
		CheckSuite struct {
			HeadBranch string `json:"head_branch"`
		} `json:"check_suite"`
	} `json:"check_run"`
	Repository github.Repository `json:"repository"`
}

func (e *CheckRun) Handle() (*discord.Webhook, error) {
	if e.Action != "completed" || e.CheckRun.Conclusion == "" {
		return nil, nil
	}

	if slices.Contains(ignoredChecks, e.CheckRun.App.Name) {
		return nil, nil
	}

	emoji := config.Get().SuccessEmoji
	if e.CheckRun.Conclusion == "failure" {
		emoji = config.Get().FailureEmoji
	}

	return &discord.Webhook{
		Content: fmt.Sprintf(
			"%s Check [%s](<%s>) on [%s](<%s>)/[%s](<%[5]s/tree/%s>)",
			emoji,
			e.CheckRun.Conclusion,
			e.CheckRun.HtmlUrl,
			e.Repository.Name,
			e.Repository.HtmlUrl,
			e.CheckRun.CheckSuite.HeadBranch,
		),
		Username:  e.CheckRun.App.Name,
		AvatarUrl: e.Repository.Owner.AvatarUrl,
	}, nil
}
